examples/idcomp.ctt

```diff
@@ -0,0 +1,253 @@
+--
+--     (p . refl)^-1  =  refl^-1 . p^-1
+--
+-- Since refl^-1 = refl definitionally the goal reduces to:
+--
+--     (p . refl)^-1  =  refl . p^-1
+--
+-- Using the primitive composition both p . refl and refl . p^-1 will
+-- reduce and leave the goal as:
+--
+--     p^-1  =  p^-1
+--
+-- which is easily proved by refl. However, using the defined
+-- composition only one side will reduce and we will need to perform
+-- another application of idJ on p. The situation where we first
+-- induct on p is symmetric.
+
+-- The lemma we want to show
+InvComp (A : U) (a b : A) (p : Id A a b) (c : A) (q : Id A b c) : U =
+  Id (Id A c a)
+    (symId A a c (idComp A a b c p q))
+    (idComp A c b a (symId A b c q) (symId A a b p))
+
+-- Proved using a single application of idJ
+invComp (A : U) (a b c : A) (p : Id A a b) (q : Id A b c) : InvComp A a b p c q =
+  idJ A b (InvComp A a b p) (refId (Id A b a) (symId A a b p)) c q
+
+-- The lemma using compId
+InvComp' (A : U) (a b : A) (p : Id A a b) (c : A) (q : Id A b c) : U =
+  Id (Id A c a)
+    (symId A a c (compId A a b c p q))
+    (compId A c b a (symId A b c q) (symId A a b p))
+
+-- Now requires two applications of idJ
+invComp' (A : U) (a b c : A) (p : Id A a b) (q : Id A b c) : InvComp' A a b p c q =
+  idJ A b (InvComp' A a b p)
+    (idJ A a
+      (\(b : A)(p : Id A a b) -> InvComp' A a b p b (refId A b))
+        (refId (Id A a a) (refId A a)) b p) c q
+
+
+--------------------------------------------------------------------------------
+-- More examples/experiements
+
+-- Weak refl
+wrefId (A : U)(a : A) : Id A a a = idC (<_> a) []
+
+s2w (A : U)(a : A) : Path (Id A a a) (refId A a) (wrefId A a)  =
+  <i> idC (<_> a) [(i=0) -> a]
+
+w2s (A : U)(a : A) : Path (Id A a a) (wrefId A a) (refId A a)  =
+  <i> idC (<_> a) [(i=1) -> a]
+
+-- We now have the following two squares:
+--   (The lines in the x-direction are regular paths,
+--    and the lines in the y-direction are Id paths)
+--
+--                 refl
+--             a -------> a
+--             |          |
+--       refId |          | wrefId
+--             v   refl   v
+--             a -------> a
+--             |          |
+--      wrefId |          | refId
+--             v   refl   v
+--             a -------> a
+--
+-- We can now compose in the vertical direction to get a square:
+--
+--                 refl
+--             a -------> a
+--  refId      |          | wrefId
+--   . wrefId  |          |   . refId
+--             v   refl   v
+--             a -------> a
+--
+-- where p . q is the composition of p followed by q.
+
+-- If . is the built-in composition then both sides are wrefId
+-- and we get the following:
+
+square (A : U)(a : A) : Path (Id A a a) (wrefId A a) (wrefId A a) =
+  <i> idComp A a a a (s2w A a @ i) (w2s A a @ i)
+
+-- Gives:
+--         <i> idC (<j> comp (<_> A) a
+--               [ (i = 0) -> <_> a
+--               , (i = 1) -> <_> a
+-- 		 , (j = 0) -> <_> a
+-- 		 , (j = 1) -> <_> a ]) []
```

Comment:
Yep, this is a tabs vs spaces issue. They look aligned in my emacs buffer! Fixed now.